LoopKit/Service.swift

```diff
@@ -45,12 +52,14 @@ public protocol Service: AnyObject {
     /// The current, serializable state of the service.
     var rawState: RawStateValue { get }
 
+    /// Is the service onboarded and ready for use?
+    var isOnboarded: Bool { get }
+
+    /// Delete the service.
+    func delete()
```

Comment:
It *is* called in the `LoopOnboarding` plugin/repo. There exists a workflow where the user first chooses to onboard the `NightscoutService` and does so (i.e. logs in), and then goes back a couple steps and then chooses *not* to onboard with `NightscoutService`. The created service should be deleted as part of the cleanup.

I can see this also being an issue with any form on onboarding where the UI allows the user to go back and "change their mind".

I'm open to ideas here, but this seemed like the cleanest solution. I could also see a `deleteViewController` on the `ServiceUI` protocol in case a UI was needed to delete/deactivate a service/CGM/pump. 

I don't think either not cleaning it up nor requiring some arbitrary "you can't go back past here" limitation in the UI is the right way to go.

Reply:
Ok, I deleted the line and the app built, but I guess I had the Tidepool Loop build selected, and not DIY.

If the user is backing out of an onboarding UI, the UI itself could detect that and clean up anything it needs to clean up. Still not sure we need this. Seems like any removal of a service/device should be by user control via a UI of that service or device, not programmatically by Loop.

If Loop does indeed need to tell a service to wipe its credentials without user involvement, then I would expect this to not be called "delete", as it's weird to send a message to an object called `delete` and then you still have the object. Maybe something like 'eraseStoredCredentials` or `logout` for services, and `discardPairing` or something similar for devices.

Reply:
How would the UI clean up anything it needs to cleanup? Once the UI has notified Loop that the service is created, Loop then knows about the service. There is currently no way for the UI to notify Loop that the service should be forgotten. That notification is all through the service implementation (currently accessed via settings) and the service's delegate (not the `setupViewController` delegate).

I suppose we could defer notifying Loop about the service until onboarding is complete. In this case, onboarding would have to be the service's delegate. That would mean that the service state would be stored inside of the onboarding state. When onboarding was fully 100% done then onboarding would notify Loop that the service was created and onboarded which would switch the service's delegate to the `ServiceManager`. I foresee significant complications when this is extended to CGMManager and PumpManager.

Or, we could show the `settingsViewController` when we wanted the user to delete it? That would be odd, I think. User first choose and sets up NightscoutService, then goes back a couple of steps, and chooses **not** to setup NightscoutService, and is then presented with the settingsViewController and has to tap the "Delete Service" button?

Or, we could add an explicit `deleteViewController` to the `ServiceUI` protocol which would show any necessary UI to delete. But what if no UI is required to delete (much like if `setupViewController` returns nil), then we'd back to a simple `delete()` func on `Service`.

Reply:
If the cgm/pump/service manager has been created, and Loop has been notified, and then the user wants to delete the manager (by backing out or something more explicit), then that user action can trigger a cgmManagerWantsDeletion type call, right? Along with any other sort of local cleanup.

Reply:
The `cgmManagerWantsDeletion` delegate callback is invoked by the `CGMManager` instance itself via the internal `notifyDelegateOfDeletion` function which is invoked by the internal `CGMManagerUI` settings UI implementation. So, for onboarding to delete the CGM manager via our current mechanisms it would have to display the settings UI and the user would have to tap "Delete CGM Manager" (or similar depending upon the CGMManager implementation).

This works similarly for pump managers and services.

Reply:
@ps2 We still need resolution on this thread. How do we want to proceed here?

Reply:
Onboarding doesn't delete the CGM Manager; if, and only if, the user has gotten to the point in the ui where we do have an actual manager (because some state needed to be saved), and the user does something *in the manager's ui* then the manager's ui should trigger a managerWantsDeletion, and do its cleanup. The onboarding plugin doesn't need external commands to "delete" the manager.

Reply:
So I'm not misunderstanding something, I see two use cases:

First use case, pump/cgm/service onboarding **cancelled**. For example:
- The user is going through onboarding and is in the "Your Devices" section.
- The user starts CGM onboarding, which pops a modal dialog.
- The user gets part way through CGM onboarding and taps the "Close" button in the upper-right corner **before** finishing the CGM onboarding.
- The CGM onboarding cleans up and deletes any created CGM manager, and then sends the completion notification (to allow the parent to dismiss the modal).

Second use case, pump/cgm/service onboarding **completed**. For example:
- The user is going through onboarding and is in the "Your Devices" section.
- The user starts CGM onboarding, which pops a modal dialog.
- The user **completes** CGM onboarding and its modal dialog is dismissed as normal.
- The user then taps the "Back" button to get **before** the CGM onboarding view that originally popped the modal dialog for the CGM onboarding. (Perhaps this is a chooser and they choose something different here.)
- How do we cleanup here?

The first use case makes sense to me. The second use case is unclear.

Note that the NightscoutOnboarding workflow allows both use cases. 

The first use case would be if they selected "Setup with Nightscout", but cancelled while entering the URL/secret.

The second use case would be if they selected "Setup with Nightscout", entered the URL/secret, completed the setup, the modal dialog dismisses, they get a few steps into the Therapy Settings workflow, then they tap "Back" a number of times until they get to the Nightscout chooser, and then they tap "Setup **without** Nightscout". How do we cleanup here?

Reply:
I think the second case is the same as normal successful completion of CGM onboarding.  As soon as CGM onboarding is complete, any way that the user exits out of the UI (either by tapping "continue" on the last page of CGM onboarding, or navigating backwards), and the main onboarding UI is shown, the main onboarding UI should not show a chooser.

Reply:
Okay, not sure if I agree here and I suspect this will need to change later on, but will do. I will look into the best way to yank the chooser UIViewController out of the UINavigationController view hierarchy after it has been displayed so that subsequent "back" operations do not encounter the chooser.

Reply:
Looks like the "delete" method is still here?

Reply:
Oops, yep, forgot to remove it now that isn't needed. Will do.